AMM/contracts/exchange/src/main.sw

```diff
@@ -0,0 +1,423 @@
+        let output_reserve_sufficient = exact_output <= output_asset_in_reserve;
+        PreviewSwapInfo {
+            amount: max_input,
+            output_reserve_sufficient: output_reserve_sufficient,
+        }
+    }
+
+    #[storage(read, write)]
+    fn remove_liquidity(min_asset_a: u64, min_asset_b: u64, deadline: u64) -> RemoveLiquidityInfo {
+        require(storage.pair.is_some(), InitError::NotInitialized);
+
+        let total_liquidity = storage.liquidity_pool_supply;
+        require(total_liquidity > 0, TransactionError::LiquidityCannotBeZero);
+
+        let (asset_a_id, asset_b_id) = storage.pair.unwrap();
+
+        require(msg_asset_id() == contract_id(), InputError::InvalidAsset);
+        require(min_asset_a > 0 && min_asset_b > 0, InputError::AmountCannotBeZero);
+        require(deadline > height(), InputError::DeadlinePassed);
+
+        let amount = msg_amount();
+
+        require(amount > 0, InputError::AmountCannotBeZero);
+
+        let sender = msg_sender().unwrap();
+        let asset_a_in_reserve = storage.reserves.get(asset_a_id);
+        let asset_b_in_reserve = storage.reserves.get(asset_b_id);
+        let asset_a_amount_to_remove = multiply_div(amount, asset_a_in_reserve, total_liquidity);
+        let asset_b_amount_to_remove = multiply_div(amount, asset_b_in_reserve, total_liquidity);
+
+        require(asset_a_amount_to_remove >= min_asset_a, TransactionError::DesiredAmountTooHigh(min_asset_a));
+        require(asset_b_amount_to_remove >= min_asset_b, TransactionError::DesiredAmountTooHigh(min_asset_b));
+
+        burn(amount);
+        storage.liquidity_pool_supply = total_liquidity - amount;
+        storage.reserves.insert(asset_b_id, asset_b_in_reserve - asset_b_amount_to_remove);
+        storage.reserves.insert(asset_a_id, asset_a_in_reserve - asset_a_amount_to_remove);
+        transfer(asset_a_amount_to_remove, asset_a_id, sender);
+        transfer(asset_b_amount_to_remove, asset_b_id, sender);
+        log(RemoveLiquidityEvent {
+            amount_a: asset_a_amount_to_remove,
+            amount_b: asset_b_amount_to_remove,
+            liquidity: amount,
+        });
+        RemoveLiquidityInfo {
+            asset_a_amount: asset_a_amount_to_remove,
+            asset_b_amount: asset_b_amount_to_remove,
+            liquidity: amount,
+        }
+    }
+
+    #[storage(read, write)]
+    fn swap_with_exact_input(min_output: Option<u64>, deadline: u64) -> u64 {
+        require(storage.pair.is_some(), InitError::NotInitialized);
+
+        let (asset_a_id, asset_b_id) = storage.pair.unwrap();
+        let input_asset = msg_asset_id();
+
+        require(input_asset == asset_a_id || input_asset == asset_b_id, InputError::InvalidAsset);
+        require(deadline >= height(), InputError::DeadlinePassed);
+
+        let exact_input = msg_amount();
+        require(exact_input > 0, InputError::AmountCannotBeZero);
+
+        let sender = msg_sender().unwrap();
```

Comment:
If possible I would remove as much of the code as you can in this function and replace it with a call to `preview_swap_with_exact_input`

Reply:
Is it possible to call a function from another function in the same contract?

Reply:
ABI functions are all external so you cannot call function `A` in contract `A` from function `B` in contract `A`

Reply:
Ah my mistake.  Ignore my comment then.